### DISCUSSION TITLE
Client Side User Registration with Directus

### DISCUSSION THREAD

**Discussion by StrangeFruit** [directus/directus#6191] [Archive]
Good Day Everyone,

I was wondering if there was a proper way to implement user registration via a client side app eg:react, vue, in a similar way that the user login is done.

I found a bit of a solution by setting the Directus System User API access to public and adding some access polices to the fields but I'm not sure if this is secure , it also  does not allow me to set roles and the password seems to automatically set to "admin".

Any assistance will be appreciated.  

Thank you.

**Comment by rijkvanzanten** [comment]
> setting the Directus System User API access to public and adding some access polices to the fields 

That's one way of doing it! By enabling create permissions, and disabling all fields except for email / password, you can let people register safely. You can use the "Presets" option to pre-fill the role to a role you deem safe for use. 

What do you mean by 

> and the password seems to automatically set to "admin".

?

**Reply by StrangeFruit** [reply]
Good Morning ,
Sorry for the delayed response , I appreciate your response and the great work you have done on this project , keep it up.

As for the password being set to admin,
these are my permissions settings:
![Screenshot 2021-06-13 at 7 10 20 AM](https://user-images.githubusercontent.com/6045532/121805545-b6ea7700-cc19-11eb-9ef8-badd09b58924.png)

and this is how I'm creating a user:
![Screenshot 2021-06-13 at 7 31 52 AM](https://user-images.githubusercontent.com/6045532/121805595-d4b7dc00-cc19-11eb-9e2e-94ebffb38617.png)

and when the user is created and I click to show the raw value, this is what i see:
![Screenshot 2021-06-13 at 7 32 32 AM](https://user-images.githubusercontent.com/6045532/121805635-f913b880-cc19-11eb-96df-051170eadcb1.png)

Also I tried your permission idea to add a specified role, is this the appropriate format ?

I got the role id from the URL of the role page (is there another way ?):
![Screenshot 2021-06-13 at 7 37 56 AM](https://user-images.githubusercontent.com/6045532/121805739-6b849880-cc1a-11eb-956f-ca6909b05833.png)

then is set the Presets to make sure it was equal to the role id : 
![Screenshot 2021-06-13 at 7 10 30 AM](https://user-images.githubusercontent.com/6045532/121805763-8ce58480-cc1a-11eb-8729-6a8ca8819f1b.png)

Let me know if any more information is needed.

**Reply by rijkvanzanten** [reply]
Heya!

Try disabling the `role` under `fields` (bit of a security risk if they can set their own role), and use the following for presets:

```json
{
  "role": "cc943 etc etc"
}
```

**Reply by StrangeFruit** [reply]
Thank you very much, that was very helpful, 

If you don't mind I have 2 specific Directus features I would like to inquire about,

1 - will the javascript sdk be adding a register user feature to its Auth function , similar to the login feature ?

2 - will the websockets API specified in the  Q3-2021 section of the Directus roadmap be integrated with Directus GraphQl feature set , specifically GraphQl subscriptions ?

**Reply by rijkvanzanten** [reply]
> 1 - will the javascript sdk be adding a register user feature to its Auth function , similar to the login feature ?

Yeah, we aim to cover everything the API can do in the JS SDK 👍🏻  The API itself will need to have a dedicated register endpoint before we can add it to the SDK tho

> 2 - will the websockets API specified in the Q3-2021 section of the Directus roadmap be integrated with Directus GraphQl feature set , specifically GraphQl subscriptions ?

Yup! That'll be real time data streams for both "REST" (through a relay WS endpoint) and GraphQL Subscriptions 👍🏻

**Reply by StrangeFruit** [reply]
Great to hear, and once again great job!!
